### PR TITLE
Simple conversion of the format returned by the API

### PR DESCRIPTION
**Pull request by ph**
## DO NOT MERGE REQUIRE CODE CHANGE IN KIBANA
---

The Kibana API for CM is undergoing internal changes to better allow
future scalability of the configurations, one of the requirements for
this move was to flatten as much as possible the configuration file so
the format would be the same independant of the key that need to
received ot the configuration. It will be up to the manager to take the
JSON format and convert it into something that our Configuration can
understand.

ref: https://github.com/elastic/kibana/pull/27717

---

Details about the implementation:

Sorry, It took much longer than I thought it would, and I initially wanted to do it more cleanly and allow the reloadable registry to tells us what it actually needed. I was actually pretty far into it but after taking a step back I was anxious about my strategy and the number of required changes and the implication in review that I've decided to go for a simple take on it instead.

Right now the registry as two concepts a `Single` and `List`, I think we should uniform them into a single `Reloable` and let the implementation details manage how a reload is happening and we also need better feedback from the reload part.

### CODE REVIEW

**Comment by ph** [comment]
@mattapperson Concerning the id that will be returned for each `config_block` will Where it will be located at the same level of the type of in the config? 

1.
```json
{
    "configuration_blocks": [

        {
            "config": {
              "sub_type": "elasticsearch"
              "hosts": [
                  "localhost"
              ],
              "password": "foobar",
              "username": "elastic"
            },
            "type": "output"
            "id": "12312341231231",
        }
    ]
}
```

2.
```
{
    "configuration_blocks": [

        {
            "config": {
              "sub_type": "elasticsearch"
              "id": "12312341231231",
              "hosts": [
                  "localhost"
              ],
              "password": "foobar",
              "username": "elastic"
            },
            "type": "output"

        }
    ]
}
```

**Comment by ph** [comment]
Winlogbeat failure on beats-ci is due to a network issue fetching the golang artifact.

**Comment by mattapperson** [comment]
`sub_type` should be `_sub_type` where the prefixed `_` ensures its known to be "special"

File: x-pack/libbeat/management/api/configuration.go [hunk 1]
```diff
@@ -50,16 +51,40 @@ func (c *ConfigBlock) ConfigWithMeta() (*reload.ConfigWithMeta, error) {
 	}, nil
 }
 
+type configResponse struct {
+	Type string
+	Raw  map[string]interface{}
+}
+
+func (c *configResponse) UnmarshalJSON(b []byte) error {
```
Comment (urso) [inline x-pack/libbeat/management/api/configuration.go]: nit: as this is a "constructor/parser" json to `*configResponse` i'd prefer this to be a function like:

```
func unmarshalConfigReponse(json []byte) (*configResponse, error) {
  ...
}
```

File: x-pack/libbeat/management/api/convert.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+// you may not use this file except in compliance with the Elastic License.
+
+package api
+
+import (
+	"fmt"
+	"strings"
+)
+
+type convertFunc func(map[string]interface{}) (map[string]interface{}, error)
```
Comment (urso) [inline x-pack/libbeat/management/api/convert.go]: s/convertFunc/converter/ ?

File: x-pack/libbeat/management/api/convert.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+func selectConverter(t string) convertFunc {
+	c := noopConvert
+	for k, v := range mapper {
+		if strings.Index(t, k) > -1 {
+			c = v
+			return c
+		}
+	}
+	return c
```
Comment (urso) [inline x-pack/libbeat/management/api/convert.go]: given the body and early return we don't really need 'c' :)

**Comment by urso** [comment]
Just from reading the code it is not really clear why the converters are needed. Maybe we should also include a doc.go file explaining the API/protocol?

**Comment by ph** [comment]
I will add a doc illustrating the before and after.

**Comment by ph** [comment]
I've enabled the tests in this PR again and also fixed the things from the review.

@mattapperson what is the value of `_sub_type` when we define a normal filebeat input?

**Comment by mattapperson** [comment]
@ph in such a case it is not defined. _sub_type is optional, and used when the value needs to be both nested and the top level of the nest needs to be dynamic.

File: x-pack/libbeat/management/api/convert.go (hunk 3 above)
Comment (urso) [reply x-pack/libbeat/management/api/convert.go]: Sorry, wasnt' clear. I was wondering if we need the variable at all:

```
for k, v := range mapper {
  if strings.Contains(t, k) {
    return v
  }
}
return noopConverter // no match
```

Comment (ph) [reply x-pack/libbeat/management/api/convert.go]: Oh yeah thats correct, I've made the change.

**Comment by ph** [comment]
This will need to go in 6.x

**Comment by codingogre** [comment]
Has the upgrade paths been tested with this change?  Upgrading from past versions to 6.7 includes this code, but looking at the .management-beats index after registering a beat looks like this:

```
{
  "_index": ".management-beats",
  "_type": "_doc",
  "_id": "tag:Filebeat-System",
  "_version": 3,
  "_score": 0,
  "_source": {
    "tag": {
      "color": "#00b3a4",
      "configuration_blocks": [
        {
          "type": "filebeat.modules",
          **"configs": [
            {
              "module": "system"
            }**
          ]
        }
      ],
      "id": "Filebeat-System",
      "last_updated": "2019-04-03T23:30:42.137Z"
    },
    "type": "tag"
  },
  "fields": {
    "tag.last_updated": [
      "2019-04-03T23:30:42.137Z"
    ]
  }
}
```

This generates errors in the filebeat log and results in filebeat not being able to pick up new configuration changes which totally breaks Centralized Beats Management.

2019-04-04T02:56:44.820-0400    ERROR   [centralmgmt]   management/manager.go:224       error retrieving new configurations, will use cached ones: error unmarshaling Kibana response: '_sub_type' key not found

Looking through the code it seems there are tests where a new config type is used:

From beats/x-pack/libbeat/management/api/configuration_test.go

```
{
	"type": "filebeat.modules",
	**"config": {
		"_sub_type": "system"
	}**
}
```

**Comment by ph** [comment]
@codingogre did you run the migration tool ? https://github.com/elastic/migrate-management-beats/

**Pull request merged**